File: samples/python/evaluation/llm_as_judge.py
from typing import TypedDict, Literal
from langchain_anthropic import ChatAnthropic
from langgraph.graph import StateGraph, END

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")

class State(TypedDict):
    task: str
    output: str
    verdict: str
    reason: str
    retries: int


def produce(state: State) -> State:
    attempt = state["retries"] + 1
    prompt = (
        f"Answer the following task concisely (attempt {attempt}/3):\n{state['task']}"
    )
    response = llm.invoke(prompt)
    print(f"[Producer] Attempt {attempt}: {response.content[:120]}")
    return {**state, "output": response.content}


def judge(state: State) -> State:
    prompt = (
        f"You are a strict evaluator. Given the task and answer below, "
        f"reply with exactly one word: APPROVED or REJECTED, then a dash and a short reason.\n\n"
        f"Task: {state['task']}\n"
        f"Answer: {state['output']}\n\n"
        f"Evaluation:"
    )
    response = llm.invoke(prompt)
    text = response.content.strip()
    verdict = "APPROVED" if text.upper().startswith("APPROVED") else "REJECTED"
    reason = text.split("-", 1)[-1].strip() if "-" in text else text
    print(f"[Judge] Verdict: {verdict} | Reason: {reason}")
    return {**state, "verdict": verdict, "reason": reason}


def route(state: State) -> Literal["produce", "dead_letter", "__end__"]:
    if state["verdict"] == "APPROVED":
        return "__end__"
    if state["retries"] < 2:
        return "produce"
    return "dead_letter"


def increment_retries(state: State) -> State:
    return {**state, "retries": state["retries"] + 1}


def dead_letter(state: State) -> State:
    print(f"[Dead Letter] Max retries reached. Last reason: {state['reason']}")
    return state


builder = StateGraph(State)
builder.add_node("produce", produce)
builder.add_node("judge", judge)
builder.add_node("increment", increment_retries)
builder.add_node("dead_letter", dead_letter)

builder.set_entry_point("produce")
builder.add_edge("produce", "judge")
builder.add_conditional_edges(
    "judge",
    route,
    {"__end__": END, "produce": "increment", "dead_letter": "dead_letter"},
)
builder.add_edge("increment", "produce")
builder.add_edge("dead_letter", END)

graph = builder.compile()


def main():
    task = "Explain what a REST API is in exactly one sentence."
    initial: State = {"task": task, "output": "", "verdict": "", "reason": "", "retries": 0}
    print(f"[Task] {task}\n")
    result = graph.invoke(initial)
    print(f"\n[Final] Verdict={result['verdict']} | Output={result['output'][:150]}")


if __name__ == "__main__":
    main()

File: samples/typescript/evaluation/llm_as_judge.ts
import { ChatAnthropic } from "@langchain/anthropic";
import { StateGraph, END, START, Annotation } from "@langchain/langgraph";

const llm = new ChatAnthropic({ model: "claude-haiku-4-5-20251001" });

const StateAnnotation = Annotation.Root({
  task: Annotation<string>(),
  output: Annotation<string>(),
  verdict: Annotation<string>(),
  retries: Annotation<number>(),
});

type State = typeof StateAnnotation.State;

async function produce(state: State): Promise<Partial<State>> {
  const res = await llm.invoke(`Complete this task concisely: ${state.task}`);
  return { output: res.content as string };
}

async function judge(state: State): Promise<Partial<State>> {
  const prompt = `Task: ${state.task}\nOutput: ${state.output}\n\nReply with exactly APPROVED or REJECTED and one sentence reason.`;
  const res = await llm.invoke(prompt);
  return { verdict: res.content as string };
}

function route(state: State): string {
  if (state.verdict.startsWith("APPROVED")) return "end";
  if (state.retries >= 2) return "dead_letter";
  return "retry";
}

async function retry(state: State): Promise<Partial<State>> {
  console.log(`Retry #${state.retries + 1}. Verdict: ${state.verdict}`);
  return { retries: state.retries + 1, output: "" };
}

async function deadLetter(state: State): Promise<Partial<State>> {
  console.log(`Dead letter after ${state.retries} retries. Last verdict: ${state.verdict}`);
  return {};
}

const graph = new StateGraph(StateAnnotation)
  .addNode("produce", produce)
  .addNode("judge", judge)
  .addNode("retry", retry)
  .addNode("dead_letter", deadLetter)
  .addEdge(START, "produce")
  .addEdge("produce", "judge")
  .addConditionalEdges("judge", route, {
    end: END,
    retry: "retry",
    dead_letter: END,
  })
  .addEdge("retry", "produce")
  .compile();

async function main() {
  const result = await graph.invoke({
    task: "Explain recursion in one sentence.",
    output: "",
    verdict: "",
    retries: 0,
  });
  console.log("Final output:", result.output);
  console.log("Final verdict:", result.verdict);
}

main().catch(console.error);
